libnopegl/src/backends/gl/pipeline_gl.c

```diff
@@ -758,6 +760,8 @@ void ngli_pipeline_gl_draw_indexed(struct pipeline *s, const struct buffer *indi
     struct pipeline_graphics *graphics = &s->graphics;
     struct program_gl *program_gl = (struct program_gl *)s->program;
 
+    s_priv->insert_memory_barriers(s);
```

Comment:
I'm sorry I still don't get it. In my understanding we had sequences like: `(draw, barrier), (draw, barrier), (dispatch, barrier), (dispatch, barrier), (draw, barrier), ...`

And now we have: `(barrier, draw, barrier), (barrier, draw, barrier), (barrier, dispatch, barrier), (barrier, dispatch, barrier), (barrier, draw, barrier), ...`

Why barrier actually helps? 🤔

Reply:
> I'm sorry I still don't get it. In my understanding we had sequences like: `(draw, barrier), (draw, barrier), (dispatch, barrier), (dispatch, barrier), (draw, barrier), ...`

A barrier is inserted after the pipeline only if the pipeline is writing to the resource. So the following scenario would cause rendering issues:

`draw SSBO (ro)`, `compute SSBO (w)`, `barrier` 

A barrier is missing between the draw and the compute, ie: write after read hazard.

> 
> And now we have: `(barrier, draw, barrier), (barrier, draw, barrier), (barrier, dispatch, barrier), (barrier, dispatch, barrier), (barrier, draw, barrier), ...`

Now we insert a barrier before and after the pipeline but only if the pipeline write to the resource.

> 
> Which barrier(s) actually help(s)? thinking

Barriers between a pipeline that reads the resource and a pipeline that writes to it:

`draw SSBO (ro)`, >> **`barrier`** <<, `compute SSBO (w)`, `barrier`